compiler/dialects/tsl.py

```diff
@@ -53,59 +53,58 @@ def get_bound_ops(
+
+            # static case
+            stride = tsl.get_stride(dim, depth)
+            if stride.bound is not None:
+                bound_op = Constant.from_int_and_width(stride.bound, IndexType())
+                result.append(bound_op)
+                result_mapping[(dim, depth)] = bound_op
+
+            # dynamic case: use the dim_op and divide by product of
+            # all lower tile sizes
+            else:
+                # get the product of all lower tile sizes
+                product_tilebounds = prod(
+                    [stride.bound for _, stride in tsl.tstrides[dim] if stride.bound]
+                )
+                div_op = Constant.from_int_and_width(product_tilebounds, IndexType())
+
+                # divide the size of the memref by the product of all lower tiles
+                bound_op = DivUI(dim_op, div_op, IndexType())
+
+                # add the ops to result
+                result.extend([div_op, bound_op])
+                result_mapping[(dim, depth)] = bound_op
+
+            # inner tile depths are all static by definition of TSL
+            for depth in range(1, tsl.tstrides[dim].depth()):
                 stride = tsl.get_stride(dim, depth)
-
-                # static case
-                if stride.bound is not None:
-                    bound_op = Constant.from_int_and_width(stride.bound, IndexType())
-                    result.append(bound_op)
-                    result_mapping[(dim, depth)] = bound_op
-
-                # dynamic case
-                # to calculate the bound of a dynamic stride,
-                # we must divide the size of the memref by the product
-                # of all lower tile sizes
-                else:
-                    # get the size of the memref
-                    if shapes:  # use shapes if supplied
-                        dim_op = shapes.pop(0)
-                    else:  # else, create own dim op to calculate shape
-                        dim_index_op = Constant.from_int_and_width(dim, IndexType())
-                        dim_op = Dim.from_source_and_index(memref, dim_index_op)
-                        result.extend([dim_index_op, dim_op])
-
-                    # get the product of all lower tile sizes
-                    product_tilebounds = prod(
-                        [
-                            stride.bound
-                            for _, stride in tsl.tstrides[dim]
-                            if stride.bound
-                        ]
-                    )
-                    div_op = Constant.from_int_and_width(
-                        product_tilebounds, IndexType()
-                    )
-
-                    # divide the size of the memref by the product of all lower tiles
-                    bound_op = DivUI(dim_op.result, div_op.result, IndexType())
-
-                    # add the ops to result
-                    result.extend([div_op, bound_op])
-                    result_mapping[(dim, depth)] = bound_op
```

Comment:
The only change here is that the outermost tile (depth = 0) is hoisted out of the for loop. This results in easier code